src/internal/collection/README.md

```diff
@@ -0,0 +1,103 @@
+
+### Triggers
+
+Two triggers exist on each postgres collection table, `update_modified_trigger` and `notify_watch_trigger`.
+
+`update_modified_trigger` will modify the row before any write to update the `updatedat` column to the current time.
+
+`notify_watch_trigger` will issue `NOTIFY` commands to one or several channels in case any clients are running watches on the collection.  For table-wide watches, the channel is straightforwardly `pwc_<table>`.  For point-watches, the trigger will also notify one channel for each index on the collection.  The channel name must depend on the value of the field being indexed so that clients do not have to discard O(n) events per interesting event.  Unfortunately, channel names must be valid postgres identifiers which puts several limitations on the channel name, so the value must be hashed.  This still leaves the possibility open for a collision in the channel name, so the client must still check that the event belongs to the watch in question (see [Listeners](#listeners) for details).
+
+The payload on the channel includes several fields:
+ * the field name of the index being notified
+ * the value of the index being notified
+ * the primary key of the row
+ * the postgres operation that caused the trigger (`'INSERT'`, `'UPDATE'`, `'DELETE'`)
+ * the timestamp of the transaction that caused the trigger
+ * a base64-encoded serialized protobuf of the item
+
+### Listeners
+
+In order to provide `watch` semantics, a postgres connection must be dedicated to listening for notifications.  This functionality is not available on `sqlx` connections, so the `pq` library is used instead.  This is the same library that provides the underlying connection for `sqlx`, although we don't get much benefit from this.  A `PostgresListener` must be passed to any collection at construction time so that watches can be performed.  The `PostgresListener` lazily connect for the first `LISTEN` operation, and all `LISTEN`s should be multiplexed through a single connection.
+
+Each `Watch` operation will listen on a postgres channel according to the watch type:
+ * `Watch`, `WatchF` - listens to the table-wide `pwc_<table>` channel
+ * `WatchOne`, `WatchOneF`, `WatchByIndex` - listens to a hashed channel `pwc_<table>_<hash>` based on the index key
+
+ Once the `LISTEN` has returned, events must be buffered until the user has read out the current state of objects in the collections.  The `Watch` operation will then issue a `Get` or `List` operation to load the current state and keep track of the last modified timestamp.  The existing items will be provided to the `Watch` as `EventPut` events, and once these are done, the buffered events will be forwarded to the `Watch`, filtering out any that arrived before loading the existing items.
```

Comment:
What is the failure mode for this? Does postgres drop from the front or back of the queue. That sort of constrains what we can do.

Reply:
I don't think postgres drops items from the queue.  Locally, if a watcher is not keeping up with its events, the channel will fill up and we will error out if the write ever blocks.  This lets us keep the listening goroutine rather tight, so it should have no problem keeping up with the listen events from postgres.